use-cases/Separate-interface-from-implementation.md

```diff
@@ -0,0 +1,29 @@
+# Separate interface from implementation
+
+## General information
+
+The interface and implementation of package definitions should be separated to ensure better maintainability, flexibility, and compatibility between different tools and approaches. A well-defined interface would allow package maintainers to focus on the implementation details while providing stability and consistency for end-users. It would also enable out-of-tree packaging libraries to interoperate with Nixpkgs more effectively, reducing the need for duplicated efforts and custom workarounds.
+
+## Concrete examples
+
+Scenario 1: Out of tree implementations (eg. lang2nix)
+Alice maintains a packaging library that simplifies python packaging. Because nixpkgs offers a well defined packaging interface, she is able to design her library so that it provides good stability without suffering from sudden breakages introduced by implementation changes in nixpkgs.
+
+Scenario 2: Change implementation details
+Bob is a maintainer of the python framework in nixpkgs. He makes an improvement on the python packaging logic. He can be confident about not breaking downstream packages because he only changed the implementation, not the interface.
+
+Scenario 3: Mixing different implementations
+Carl wants to build a python project with dependencies. Some of the dependencies are defined by a lock file and are not available from nixpkgs. Carl would like to build some of the packages using poetry2nix while mixing in other existing package from nixpkgs seamlessly.
+
```

Comment:
```suggestion

Scenario 4: Upgrading a package
Dave does some refactoring to the Emacs Nix expression in Nixpkgs, but it's reverted because it breaks things for some users who were doing some `.override*` on Emacs. Dave would like to know what parts of the Emacs Nix expression should be considered an "interface".
```
(Real-world example where `s/Dave/Eelco`, https://github.com/NixOS/nixpkgs/commit/14154fea5263b42b8ad982d7efe2061201f60f70.)